std/algorithm/sorting.d

```diff
@@ -3386,17 +3386,24 @@ auto topN(alias less = "a < b",
         SwapStrategy ss = SwapStrategy.unstable,
         Range1, Range2)(Range1 r1, Range2 r2)
     if (isRandomAccessRange!(Range1) && hasLength!Range1 &&
-            isInputRange!Range2 && is(ElementType!Range1 == ElementType!Range2))
+            isInputRange!Range2 && is(ElementType!Range1 == ElementType!Range2) &&
+            hasLvalueElements!Range1 && hasLvalueElements!Range2)
 {
     import std.container : BinaryHeap;
 
     static assert(ss == SwapStrategy.unstable,
             "Stable topN not yet implemented");
     auto heap = BinaryHeap!(Range1, less)(r1);
-    for (; !r2.empty; r2.popFront())
+
+    foreach (ref e; r2)
     {
-        heap.conditionalInsert(r2.front);
+        auto hf = heap.front;
+        if (heap.conditionalInsert(e))
+        {
+            e = hf;
+        }
```

Comment:
OK I see what you're doing here. This works but it's inefficient because it does a bunch of assignments, many of which are unnecessary. The root cause seems to be the API `conditionalInsert`. How about we do this - add this method to BinaryHeap:

```
bool conditionalSwap(ref ElementType!Store value);
```

The `ref` is important here. The method will conditionally swap `value` into the heap, and will swap the "lost" element of the heap into `value`. Then this loop only needs to run

```
foreach (ref e; r2)
{
    heap.conditionalSwap(e);
}
```